### PR TITLE
feat: support aggregate function for vector.

### LINKED ISSUE #458: feat: Support vector aggregation function
Function to be implemented:
Aggregation:
- avg(vector)
- sum(vector)

Function:
- vector_norm(vector)
- vector_dims(vector)

### PR DESCRIPTION
**Pull request by my-vegetable-has-exploded**
close #458 

This pr mainly contains function including `avg` ，`sum`, `vector_dims`, `vector_norm`.

### vector_dims(vector)

Returns the dimensions of a given vector.

Arguments：a vector.
Return type integer representing the number of dimensions of the given vector.

### vector_norms(vector)

Calculates the Euclidean norm of a given vector.
Arguments:a vector.
Return double precision representing the Euclidean norm of the given vector.

### AVG

Calculates the average of the processed vectors.
Arguments
vector:  vectors.
Return vector representing the average of processed vectors.

### SUM
Arguments： vectors.
Return type vector representing the sum of processed vectors.

The aggregate function support need to add a state transition function sfunc and an optional final calculation function ffunc, following [create aggregate document](https://www.postgresql.org/docs/current/sql-createaggregate.html).  And we add `COMBINEFUNC`  here to support parallel aggregate.

### CODE REVIEW

**Comment by my-vegetable-has-exploded** [comment]
I will update the documentation for these functions, as well as the documentation for veci8 later.

File: src/datatype/functions_vecf32.rs [hunk 1]
```diff
@@ -0,0 +1,125 @@
+use crate::datatype::memory_vecf32::{Vecf32Input, Vecf32Output};
+use crate::error::*;
+use base::scalar::*;
+use base::vector::*;
+
+/// accumulate intermediate state for vector average
+#[pgrx::pg_extern(immutable, strict, parallel_safe)]
+fn _vectors_vector_accum(
+    mut state: pgrx::composite_type!('static, "vectors.vector_accum_state"),
```
Comment (VoVAllen) [inline src/datatype/functions_vecf32.rs:9]: Is the schema name hardcoded now? @usamoi

File: src/sql/finalize.sql [hunk 2]
```diff
@@ -593,6 +598,42 @@ $$;
...
 STRICT LANGUAGE c AS 'MODULE_PATHNAME', '_vectors_alter_vector_index_wrapper';
 
+CREATE FUNCTION vector_dims("v" vector) RETURNS INT
+STRICT PARALLEL SAFE LANGUAGE c AS 'MODULE_PATHNAME', '_vectors_vector_dims_wrapper';
+
+CREATE FUNCTION vector_norm("v" vector) RETURNS real
+STRICT PARALLEL SAFE LANGUAGE c AS 'MODULE_PATHNAME', '_vectors_vector_norm_wrapper';
+
+CREATE FUNCTION vector_add("v1" vector, "v2" vector) RETURNS vector 
+STRICT PARALLEL SAFE LANGUAGE c AS 'MODULE_PATHNAME', '_vectors_vecf32_operator_add_wrapper';
```
Comment (usamoi) [inline src/sql/finalize.sql:608]: Why don't we just use `+`?

File: src/sql/finalize.sql [hunk 3]
```diff
@@ -593,6 +598,42 @@ $$;
...
+
+CREATE FUNCTION vector_accum("state" vector_accum_state, "value" vector) RETURNS vector_accum_state	
+STRICT PARALLEL SAFE LANGUAGE c AS 'MODULE_PATHNAME', '_vectors_vector_accum_wrapper';
+
+CREATE FUNCTION vector_combine("state1" vector_accum_state, "state2" vector_accum_state) RETURNS vector_accum_state
+STRICT PARALLEL SAFE LANGUAGE c AS 'MODULE_PATHNAME', '_vectors_vector_combine_wrapper';
+
+CREATE FUNCTION vector_final("state" vector_accum_state) RETURNS vector
+STRICT PARALLEL SAFE LANGUAGE c AS 'MODULE_PATHNAME', '_vectors_vector_final_wrapper';
```
Comment (usamoi) [inline src/sql/finalize.sql:617]: Suggest not exposing these functions to users.

File: src/datatype/functions_vecf32.rs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        state.set_by_name("sum", sum).unwrap();
+        state
+    }
+}
+
+#[pgrx::pg_extern(immutable, strict, parallel_safe)]
+fn _vectors_vector_combine(
+    state1: pgrx::composite_type!('static, "vectors.vector_accum_state"),
+    state2: pgrx::composite_type!('static, "vectors.vector_accum_state"),
+) -> pgrx::composite_type!('static, "vectors.vector_accum_state") {
```
Comment (usamoi) [inline src/datatype/functions_vecf32.rs:47]: `pgrx::composite_type!('a, ...)` should be better.

File: src/sql/finalize.sql [hunk 5]
```diff
@@ -78,6 +78,11 @@ CREATE TYPE vector_index_stat AS (
     idx_options TEXT
 );
 
+CREATE TYPE vector_accum_state AS (
+	count BIGINT,
+	sum double precision[]
+);
```
Comment (usamoi) [inline src/sql/finalize.sql:84]: I feel uneasy about using a composite type as an aggregate state. `set_by_name` calls PostgreSQL functions so there are `sigsetjmp` cost. I'm not sure if it matters.

File: src/sql/finalize.sql (hunk 2 above)
Comment (my-vegetable-has-exploded) [reply src/sql/finalize.sql:608]: Such sql don't works for me. Is there any better way?
```
CREATE AGGREGATE sum(vector) (
	SFUNC = +,
	STYPE = vector,
	COMBINEFUNC = +,
	PARALLEL = SAFE
);
```

File: src/sql/finalize.sql (hunk 5 above)
Comment (my-vegetable-has-exploded) [reply src/sql/finalize.sql:84]: I also think that it is a better way to ceate a new type and implement `IntoDatum` and `From Datum`. The cost will be less this way.

File: src/sql/finalize.sql (hunk 2 above)
Comment (usamoi) [reply src/sql/finalize.sql:608]:
```
CREATE AGGREGATE sum(vector) (
        SFUNC = _vectors_vecf32_operator_add,
        STYPE = vector,
        COMBINEFUNC = _vectors_vecf32_operator_add,
        PARALLEL = SAFE
);
```

It works for me.

Comment (my-vegetable-has-exploded) [reply src/sql/finalize.sql:608]: get! thanks

**Comment by usamoi** [comment]
Suggest add `vector_l2_norm` here. It helps users to take advantage of dot product as distance function.

**Comment by my-vegetable-has-exploded** [comment]
> Suggest add `vector_l2_norm` here. It helps users to take advantage of dot product as distance function.

Is this same as `vector_norm` ?

**Comment by usamoi** [comment]
> > Suggest add `vector_l2_norm` here. It helps users to take advantage of dot product as distance function.
> 
> Is this same as `vector_norm` ?

#255

File: src/sql/finalize.sql (hunk 3 above)
Comment (my-vegetable-has-exploded) [reply src/sql/finalize.sql:617]: I misunderstood the purpose of `CREATE FUNCTION`.

File: src/datatype/functions_vecf32.rs [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        result.set_by_name("count", count + 1).unwrap();
+        result.set_by_name("sum", sum).unwrap();
+        result
+    } else {
+        let sum = state
+            .get_by_name::<pgrx::Array<f64>>("sum")
+            .unwrap()
+            .unwrap();
+        check_matched_dims(sum.len(), value.dims());
+        // TODO: pgrx::Array<T> don't support mutable operations currently, we can reuse the state once it's supported.
```
Comment (my-vegetable-has-exploded) [inline src/datatype/functions_vecf32.rs:31]: I double check this solution. It seem that the memory management of `AggState` is well-designed. For `transfn` like `sfunc( internal-state, next-data-values ) ---> next-internal-state` , return `internal-state` after modification as result is just okay.
See https://github.com/postgres/postgres/blob/52b49b796cc7fd976f4da6aa49c9679ecdae8bd5/src/backend/executor/nodeAgg.c#L81-L100 and https://doxygen.postgresql.org/int8_8c.html#aff6c10f1da41c3aca8253c3e1d6f1378 for more details.
I am going to change the `composite_type` to a struct. And modify the state in-place.

File: src/datatype/functions_vecf32.rs [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+    for &x in state.slice().iter().skip(1) {
+        buffer.push_str(format!(", {}", x).as_str());
+    }
+    buffer.push(']');
+    CString::new(buffer).unwrap()
+}
+
+/// accumulate intermediate state for vector average
+#[pgrx::pg_extern(immutable, strict, parallel_safe)]
+fn _vectors_vector_accum<'a>(
```
Comment (my-vegetable-has-exploded) [inline src/datatype/functions_vecf32.rs:209]: I meet unused-lifetimes warning here. I don't configure out the reason here. As far as I know , this warning only occurs with #[pgrx::pg_extern].
The 235th line is the same.

Comment (usamoi) [reply src/datatype/functions_vecf32.rs:209]: The macro expands to
```rust
pub unsafe extern "C" fn _vectors_bvecf32_send_wrapper<'a>(
    _fcinfo: ::pgrx::pg_sys::FunctionCallInfo,
) -> ::pgrx::pg_sys::Datum
```
The lifetime parameter is redundant here. You could use `allow` to suppress this warning before pgrx fixes it.

Comment (my-vegetable-has-exploded) [reply src/datatype/functions_vecf32.rs:209]: I just discovered this problem too. After add `#![allow(unused_lifetimes)] #![allow(clippy::extra_unused_lifetimes)]` in file header, it works. Thanks!

File: src/datatype/functions_vecf32.rs [hunk 8]
```diff
@@ -0,0 +1,292 @@
+#![allow(unused_lifetimes)]
```
Comment (my-vegetable-has-exploded) [inline src/datatype/functions_vecf32.rs:1]: relate https://github.com/pgcentralfoundation/pgrx/issues/1649

**Comment by VoVAllen** [comment]
@usamoi PTAL

File: src/datatype/functions_vecf32.rs [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+/// Get the dimensions of a vector.
+#[pgrx::pg_extern(immutable, strict, parallel_safe)]
+fn _vectors_vector_dims(vector: Vecf32Input<'_>) -> i32 {
+    vector.dims() as i32
+}
+
+/// Calculate the l2 norm of a vector.
+#[pgrx::pg_extern(immutable, strict, parallel_safe)]
+fn _vectors_vector_norm(vector: Vecf32Input<'_>) -> f32 {
+    (-Vecf32Dot::distance(vector.for_borrow(), vector.for_borrow()).to_f32()).sqrt()
```
Comment (usamoi) [inline src/datatype/functions_vecf32.rs:289]: Suggest adding `l2_norm` to `VectorBorrowed`. `-` here looks strange.

File: src/sql/bootstrap.sql [hunk 10]
```diff
@@ -8,5 +8,6 @@ CREATE TYPE svector;
 CREATE TYPE bvector;
 CREATE TYPE veci8;
 CREATE TYPE vector_index_stat;
+CREATE TYPE accumulate_state;
```
Comment (usamoi) [inline src/sql/bootstrap.sql:11]: Suggest renaming to `vector_accumulate_state`.

File: src/datatype/functions_vecf32.rs [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+use pgrx::pgrx_sql_entity_graph::metadata::SqlMapping;
+use pgrx::pgrx_sql_entity_graph::metadata::SqlTranslatable;
+use pgrx::{FromDatum, IntoDatum};
+use std::alloc::Layout;
+use std::ffi::{CStr, CString};
+use std::ops::{Deref, DerefMut};
+use std::ptr::NonNull;
+
+#[repr(C, align(8))]
+pub struct AccumulateStateHeader {
```
Comment (usamoi) [inline src/datatype/functions_vecf32.rs:22]: It'll be better if it's more generic so that we use reuse it for more usage.

Comment (my-vegetable-has-exploded) [reply src/datatype/functions_vecf32.rs:21]: I think only `AccumulateState` for `vecf32` and `AccumulateState` for `vecf16` layout can be same?  Just adding a generic type seems won't be so helpful?

**Comment by VoVAllen** [comment]
@usamoi @my-vegetable-has-exploded Is this ready to merge?

**Review by usamoi — APPROVED** [review]

**Comment by my-vegetable-has-exploded** [comment]
> @usamoi @my-vegetable-has-exploded Is this ready to merge?

I think it's ready to go now.

**Pull request merged**